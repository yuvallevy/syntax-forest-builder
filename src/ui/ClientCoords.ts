type ClientX = number;
type ClientY = number;

type ClientCoords = {
  clientX: ClientX;
  clientY: ClientY;
};

export type ClientRect = {
  topLeft: ClientCoords,
  bottomRight: ClientCoords,
};

export default ClientCoords;
