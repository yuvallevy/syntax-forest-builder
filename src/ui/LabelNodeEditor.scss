.LabelNodeEditorInput {
  position: absolute;
  width: 2em;
  border: 1px solid #ccc;
  border-radius: 3px;
  font-family: 'Times New Roman', Times, serif;
  font-size: 16px;
  text-align: center;
  transform: translate(-50%, -100%);
}
