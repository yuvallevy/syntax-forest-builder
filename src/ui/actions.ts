import { union } from '../core/objTransforms';
import { Id } from '../core/types';
import { UiState } from './state';

type UiAction =
  | { type: 'selectNodes', plotId: Id, treeIds: Id[], nodeIds: Id[], mode: 'set' | 'add' }
;

export const reducer = (state: UiState, action: UiAction): UiState => {
  switch (action.type) {
    case 'selectNodes':
      return {
        ...state,
        activePlotId: action.plotId,
        selectedTreeIds: action.mode === 'add' ? union(state.selectedTreeIds, action.treeIds) : action.treeIds,
        selectedNodeIds: action.mode === 'add' ? union(state.selectedNodeIds, action.nodeIds) : action.nodeIds,
      };
    default:
      return state;
  }
};
