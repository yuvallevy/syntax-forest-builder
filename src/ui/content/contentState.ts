import {
  Id, StringSlice, Sentence, NodeIndicatorInPlot, PlotIndex
} from '../../content/types';
import {
  adoptNodesInTree, disownNodesInTree,
  InsertedNode, insertNodeIntoTree, transformNodeInTree, transformNodesInTree
} from '../../content/unpositioned/manipulation';
import { deleteNodesInPlot, transformNodesInPlot } from '../../content/unpositioned/plotManipulation';
import UndoRedoHistory, { ApplyActionFunc, applyToHistory, redo, ReverseActionFunc, undo, UndoableActionCommon } from '../../util/UndoRedoHistory';
import { handleLocalSentenceChange } from './editNodes';
import { changeAt, insertAt, omitKey, removeAt } from '../../util/objTransforms';
import {
  isTerminal, PlotCoordsOffset, UnpositionedPlot, UnpositionedTree
} from '../../content/unpositioned/types';

/**
 * Represents an action taken by the user.
 * To integrate with the undo/redo system, each action is translated into an undoable "state change",
 * which is subsequently applied in a reversible fashion.
 */
export type ContentAction =
  | { type: 'addPlot' }
  | { type: 'deletePlot', plotIndex: PlotIndex }
  | { type: 'insertNode', plotIndex: PlotIndex, treeId: Id, newNodeId: Id, newNode: InsertedNode }
  | { type: 'deleteNodes', plotIndex: PlotIndex, nodeIndicators: NodeIndicatorInPlot[] }
  | { type: 'adoptNodes', plotIndex: PlotIndex, treeId: Id, adoptingNodeId: Id, adoptedNodeIds: Id[] }
  | { type: 'disownNodes', plotIndex: PlotIndex, treeId: Id, disowningNodeId: Id, disownedNodeIds: Id[] }
  | { type: 'moveNodes', plotIndex: PlotIndex, nodeIndicators: NodeIndicatorInPlot[], dx: number, dy: number }
  | { type: 'resetNodePositions', plotIndex: PlotIndex, nodeIndicators: NodeIndicatorInPlot[] }
  | { type: 'setNodeLabel', plotIndex: PlotIndex, nodeIndicator: NodeIndicatorInPlot, newLabel: string }
  | { type: 'setTriangle', plotIndex: PlotIndex, nodeIndicators: NodeIndicatorInPlot[], triangle: boolean }
  | { type: 'setSentence', plotIndex: PlotIndex, treeId: Id, newSentence: Sentence, oldSelectedSlice: StringSlice }
  | { type: 'addTree', plotIndex: PlotIndex, newTreeId: Id, offset: PlotCoordsOffset }
  | { type: 'removeTree', plotIndex: PlotIndex, treeId: Id }
;

/**
 * Represents a change in state as a result of an action done by the user.
 * Each state change includes information about the state before it so it can be easily undone,
 * and each action by the user is translated into a state change so that undo/redo can work smoothly.
 */
type ContentChange = (
  | { type: 'setPlot', plotIndex: PlotIndex, old: UnpositionedPlot, new: UnpositionedPlot }
  | { type: 'addPlot', newPlotIndex: PlotIndex, newPlot: UnpositionedPlot }
  | { type: 'removePlot', plotIndex: PlotIndex, removedPlot: UnpositionedPlot }
  | { type: 'setTree', plotIndex: PlotIndex, treeId: Id, old: UnpositionedTree, new: UnpositionedTree }
  | { type: 'addTree', plotIndex: PlotIndex, newTreeId: Id, newTree: UnpositionedTree }
  | { type: 'removeTree', plotIndex: PlotIndex, treeId: Id, removedTree: UnpositionedTree }
);

type UndoableContentChange = UndoableActionCommon & ContentChange;

/**
 * Translates a user action into a state change, which can later be undone.
 */
const makeUndoable = (state: ContentState) => (action: ContentAction): ContentChange => {
  switch (action.type) {
    case 'addPlot': {
      return {
        type: 'addPlot',
        newPlotIndex: state.plots.length,
        newPlot: { trees: {} },
      };
    }
    case 'deletePlot': {
      return {
        type: 'removePlot',
        plotIndex: action.plotIndex,
        removedPlot: state.plots[action.plotIndex],
      }
    }
    case 'insertNode': {
      return {
        type: 'setTree',
        plotIndex: action.plotIndex,
        treeId: action.treeId,
        old: state.plots[action.plotIndex].trees[action.treeId],
        new: insertNodeIntoTree(action.newNode)(action.newNodeId)(state.plots[action.plotIndex].trees[action.treeId]),
      };
    }
    case 'deleteNodes': {
      return {
        type: 'setPlot',
        plotIndex: action.plotIndex,
        old: state.plots[action.plotIndex],
        new: deleteNodesInPlot(action.nodeIndicators)(state.plots[action.plotIndex]),
      };
    }
    case 'adoptNodes': {
      return {
        type: 'setTree',
        plotIndex: action.plotIndex,
        treeId: action.treeId,
        old: state.plots[action.plotIndex].trees[action.treeId],
        new: adoptNodesInTree(action.adoptingNodeId, action.adoptedNodeIds)(
          state.plots[action.plotIndex].trees[action.treeId]),
      };
    }
    case 'disownNodes': {
      return {
        type: 'setTree',
        plotIndex: action.plotIndex,
        treeId: action.treeId,
        old: state.plots[action.plotIndex].trees[action.treeId],
        new: disownNodesInTree(action.disowningNodeId, action.disownedNodeIds)(
          state.plots[action.plotIndex].trees[action.treeId]),
      };
    }
    case 'moveNodes': {
      return {
        type: 'setPlot',
        plotIndex: action.plotIndex,
        old: state.plots[action.plotIndex],
        new: transformNodesInPlot(node => ({
          ...node,
          offset: { dTreeX: node.offset.dTreeX + action.dx, dTreeY: node.offset.dTreeY + action.dy }
        }))(action.nodeIndicators)(state.plots[action.plotIndex]),
      }
    }
    case 'resetNodePositions': {
      return {
        type: 'setPlot',
        plotIndex: action.plotIndex,
        old: state.plots[action.plotIndex],
        new: transformNodesInPlot(node => ({ ...node, offset: { dTreeX: 0, dTreeY: 0 } }))(action.nodeIndicators)(
          state.plots[action.plotIndex]),
      };
    }
    case 'setNodeLabel': {
      return {
        type: 'setTree',
        plotIndex: action.plotIndex,
        treeId: action.nodeIndicator.treeId,
        old: state.plots[action.plotIndex].trees[action.nodeIndicator.treeId],
        new: transformNodeInTree(node => ({ ...node, label: action.newLabel }))(action.nodeIndicator.nodeId)(
          state.plots[action.plotIndex].trees[action.nodeIndicator.treeId]),
      };
    }
    case 'setTriangle': {
      const treeId = action.nodeIndicators[0].treeId;  // TODO: Use all trees
      const nodeIds = action.nodeIndicators
        .filter(nodeIndicator => nodeIndicator.treeId === treeId).map(nodeIndicator => nodeIndicator.nodeId);
      return {
        type: 'setTree',
        plotIndex: action.plotIndex,
        treeId,
        old: state.plots[action.plotIndex].trees[treeId],
        new: transformNodesInTree(node => isTerminal(node) ? { ...node, triangle: action.triangle } : node)(nodeIds)(
          state.plots[action.plotIndex].trees[treeId]),
      };
    }
    case 'setSentence': {
      return {
        type: 'setTree',
        plotIndex: action.plotIndex,
        treeId: action.treeId,
        old: state.plots[action.plotIndex].trees[action.treeId],
        new: handleLocalSentenceChange(action.newSentence, action.oldSelectedSlice)(
          state.plots[action.plotIndex].trees[action.treeId]),
      };
    }
    case 'addTree': {
      return {
        type: 'addTree',
        plotIndex: action.plotIndex,
        newTreeId: action.newTreeId,
        newTree: {
          nodes: {},
          sentence: '',
          offset: action.offset,
        },
      };
    }
    case 'removeTree': {
      return {
        ...action,
        removedTree: state.plots[action.plotIndex].trees[action.treeId],
      };
    }
  }
};

const applyUndoableAction: ApplyActionFunc<UndoableContentChange, ContentState> = action => state => {
  switch (action.type) {
    case 'setPlot': {
      return {
        ...state,
        plots: changeAt(state.plots, action.plotIndex, action.new),
      };
    }
    case 'addPlot': {
      return {
        ...state,
        plots: insertAt(state.plots, action.newPlotIndex, action.newPlot),
      };
    }
    case 'removePlot': {
      return {
        ...state,
        plots: removeAt(state.plots, action.plotIndex),
      }
    }
    case 'setTree': {
      return {
        ...state,
        plots: changeAt(state.plots, action.plotIndex, {
          ...state.plots[action.plotIndex],
          trees: {
            ...state.plots[action.plotIndex].trees,
            [action.treeId]: action.new,
          },
        }),
      };
    }
    case 'addTree': {
      return {
        ...state,
        plots: changeAt(state.plots, action.plotIndex, {
          ...state.plots[action.plotIndex],
          trees: {
            ...state.plots[action.plotIndex].trees,
            [action.newTreeId]: action.newTree,
          },
        }),
      };
    }
    case 'removeTree': {
      return {
        ...state,
        plots: changeAt(state.plots, action.plotIndex, {
          ...state.plots[action.plotIndex],
          trees: omitKey(state.plots[action.plotIndex].trees, action.treeId)
        }),
      };
    }
    default:
      return state;
  }
};

const reverseUndoableAction: ReverseActionFunc<UndoableContentChange> = action => {
  switch (action.type) {
    case 'setPlot': {
      return {
        ...action,
        old: action.new,
        new: action.old,
      };
    }
    case 'addPlot': {
      return {
        ...action,
        type: 'removePlot',
        plotIndex: action.newPlotIndex,
        removedPlot: action.newPlot || { trees: {} },
      };
    }
    case 'removePlot': {
      return {
        ...action,
        type: 'addPlot',
        newPlotIndex: action.plotIndex,
        newPlot: action.removedPlot,
      }
    }
    case 'setTree': {
      return {
        ...action,
        old: action.new,
        new: action.old,
      };
    }
    case 'addTree': {
      return {
        ...action,
        type: 'removeTree',
        plotIndex: action.plotIndex,
        treeId: action.newTreeId,
        removedTree: action.newTree,
      };
    }
    case 'removeTree': {
      return {
        ...action,
        type: 'addTree',
        plotIndex: action.plotIndex,
        newTreeId: action.treeId,
        newTree: action.removedTree,
      };
    }
    default:
      return action;
  }
};

export type ContentState = {
  plots: UnpositionedPlot[];
};

export type UndoableContentState = UndoRedoHistory<UndoableContentChange, ContentState>;

const initialState: ContentState = {
  plots: [
    {
      trees: {},
    },
  ],
};

export const initialContentState: UndoableContentState = {
  current: initialState,
  undoStack: [],
  redoStack: [],
};

export const contentReducer = (state: UndoableContentState, action: ContentAction | { type: 'undo' } | { type: 'redo' }): UndoableContentState =>
  action.type === 'undo' ? undo(applyUndoableAction)(reverseUndoableAction)(state)
    : action.type === 'redo' ? redo(applyUndoableAction)(state)
    : applyToHistory(applyUndoableAction)({ ...makeUndoable(state.current)(action), timestamp: new Date() })(state);
