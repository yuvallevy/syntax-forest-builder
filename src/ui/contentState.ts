import {
  Id, IdMap, StringSlice, Sentence, UnpositionedPlot, UnpositionedTree, TreeAndNodeId, PlotCoordsOffset
} from '../core/types';
import { deleteNodesInTree, InsertedNode, insertNodeIntoTree, transformNodeInTree } from '../mantle/manipulation';
import UndoRedoHistory, { ApplyActionFunc, applyToHistory, redo, ReverseActionFunc, undo, UndoableActionCommon } from '../mantle/UndoRedoHistory';
import { handleLocalSentenceChange } from './editNodes';
import { omitKey } from '../core/objTransforms';

/**
 * Represents an action taken by the user.
 * To integrate with the undo/redo system, each action is translated into an undoable "state change",
 * which is subsequently applied in a reversible fashion.
 */
export type ContentAction =
  | { type: 'insertNode', plotId: Id, treeId: Id, newNodeId: Id, newNode: InsertedNode }
  | { type: 'deleteNodes', plotId: Id, nodes: TreeAndNodeId[] }
  | { type: 'setNodeLabel', plotId: Id, node: TreeAndNodeId, newLabel: string }
  | { type: 'setSentence', plotId: Id, treeId: Id, newSentence: Sentence, oldSelectedSlice: StringSlice }
  | { type: 'addTree', plotId: Id, newTreeId: Id, offset: PlotCoordsOffset }
  | { type: 'removeTree', plotId: Id, treeId: Id }
;

/**
 * Represents a change in state as a result of an action done by the user.
 * Each state change includes information about the state before it so it can be easily undone,
 * and each action by the user is translated into a state change so that undo/redo can work smoothly.
 */
type ContentChange = (
  | { type: 'setTree', plotId: Id, treeId: Id, old: UnpositionedTree, new: UnpositionedTree }
  | { type: 'addTree', plotId: Id, newTreeId: Id, newTree: UnpositionedTree }
  | { type: 'removeTree', plotId: Id, treeId: Id, removedTree: UnpositionedTree }
);

type UndoableContentChange = UndoableActionCommon & ContentChange;

/**
 * Translates a user action into a state change, which can later be undone.
 */
const makeUndoable = (state: ContentState) => (action: ContentAction): ContentChange => {
  switch (action.type) {
    case 'insertNode':
      return {
        type: 'setTree',
        plotId: action.plotId,
        treeId: action.treeId,
        old: state.plots[action.plotId].trees[action.treeId],
        new: insertNodeIntoTree(action.newNode)(action.newNodeId)(state.plots[action.plotId].trees[action.treeId]),
      };
    case 'deleteNodes':
      const treeId = action.nodes[0].treeId;  // TODO: Use all trees
      return {
        type: 'setTree',
        plotId: action.plotId,
        treeId: treeId,
        old: state.plots[action.plotId].trees[treeId],
        new: deleteNodesInTree(action.nodes.map(({ nodeId }) => nodeId))(state.plots[action.plotId].trees[treeId]),
      };
    case 'setNodeLabel':
      return {
        type: 'setTree',
        plotId: action.plotId,
        treeId: action.node.treeId,
        old: state.plots[action.plotId].trees[action.node.treeId],
        new: transformNodeInTree(node => ({ ...node, label: action.newLabel }))(action.node.nodeId)(
          state.plots[action.plotId].trees[action.node.treeId]),
      };
    case 'setSentence':
      return {
        type: 'setTree',
        plotId: action.plotId,
        treeId: action.treeId,
        old: state.plots[action.plotId].trees[action.treeId],
        new: handleLocalSentenceChange(action.newSentence, action.oldSelectedSlice)(
          state.plots[action.plotId].trees[action.treeId]),
      };
    case 'addTree':
      return {
        type: 'addTree',
        plotId: action.plotId,
        newTreeId: action.newTreeId,
        newTree: {
          nodes: {},
          sentence: '',
          offset: action.offset,
        },
      };
    case 'removeTree':
      return {
        ...action,
        removedTree: state.plots[action.plotId].trees[action.treeId],
      };
  }
};

const applyUndoableAction: ApplyActionFunc<UndoableContentChange, ContentState> = action => state => {
  switch (action.type) {
    case 'setTree':
      return {
        ...state,
        plots: {
          ...state.plots,
          [action.plotId]: {
            ...state.plots[action.plotId],
            trees: {
              ...state.plots[action.plotId].trees,
              [action.treeId]: action.new,
            },
          },
        },
      };
    case 'addTree':
      return {
        ...state,
        plots: {
          ...state.plots,
          [action.plotId]: {
            ...state.plots[action.plotId],
            trees: {
              ...state.plots[action.plotId].trees,
              [action.newTreeId]: action.newTree,
            },
          },
        },
      };
    case 'removeTree':
      return {
        ...state,
        plots: {
          ...state.plots,
          [action.plotId]: {
            ...state.plots[action.plotId],
            trees: omitKey(state.plots[action.plotId].trees, action.treeId)
          },
        },
      };
    default:
      return state;
  }
};

const reverseUndoableAction: ReverseActionFunc<UndoableContentChange> = action => {
  switch (action.type) {
    case 'setTree':
      return {
        ...action,
        old: action.new,
        new: action.old,
      };
    case 'addTree':
      return {
        ...action,
        type: 'removeTree',
        plotId: action.plotId,
        treeId: action.newTreeId,
        removedTree: action.newTree,
      };
    case 'removeTree':
      return {
        ...action,
        type: 'addTree',
        plotId: action.plotId,
        newTreeId: action.treeId,
        newTree: action.removedTree,
      };
    default:
      return action;
  }
};

export type ContentState = {
  plots: IdMap<UnpositionedPlot>;
};

export type UndoableContentState = UndoRedoHistory<UndoableContentChange, ContentState>;

const initialState: ContentState = {
  plots: {
    'plot': {
      trees: {},
    },
  },
};

export const initialContentState: UndoableContentState = {
  current: initialState,
  undoStack: [],
  redoStack: [],
};

export const contentReducer = (state: UndoableContentState, action: ContentAction | { type: 'undo' } | { type: 'redo' }): UndoableContentState =>
  action.type === 'undo' ? undo(applyUndoableAction)(reverseUndoableAction)(state)
    : action.type === 'redo' ? redo(applyUndoableAction)(reverseUndoableAction)(state)
    : applyToHistory(applyUndoableAction)({ ...makeUndoable(state.current)(action), timestamp: new Date() })(state);
