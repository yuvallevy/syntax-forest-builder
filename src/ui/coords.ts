import { PlotCoords, PositionedNode, PositionedTree } from '../content/positioned/types';

type ClientX = number;
type ClientY = number;

export type ClientCoords = {
  clientX: ClientX;
  clientY: ClientY;
};

export type ClientRect = {
  topLeft: ClientCoords;
  bottomRight: ClientCoords;
};

export type PlotRect = {
  topLeft: PlotCoords;
  bottomRight: PlotCoords;
};

// Calculate plot coords from client coords & vice versa, when they're different (when zoom and/or pan are implemented)

export const plotCoordsToClientCoords = (plotCoords: PlotCoords): ClientCoords => ({
  clientX: plotCoords.plotX,
  clientY: plotCoords.plotY,
});

export const clientCoordsToPlotCoords = (clientCoords: ClientCoords): PlotCoords => ({
  plotX: clientCoords.clientX,
  plotY: clientCoords.clientY,
});

export const clientRectToPlotRect = (clientRect: ClientRect): PlotRect => ({
  topLeft: clientCoordsToPlotCoords(clientRect.topLeft),
  bottomRight: clientCoordsToPlotCoords(clientRect.bottomRight),
});

/**
 * Returns the center coordinate of the given node in the given tree with respect to its containing plot.
 */
export const calculateNodeCenterOnPlot = (tree: PositionedTree) => (node: PositionedNode): PlotCoords => ({
  plotX: tree.position.plotX + node.position.treeX,
  plotY: tree.position.plotY + node.position.treeY - 9,
});
