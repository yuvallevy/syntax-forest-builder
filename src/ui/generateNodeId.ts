const ID_TIME_LENGTH = 6;
const ID_RANDOM_LENGTH = 4;
const ID_BASE = 36;

const maxTimestamp = parseInt('1' + '0'.repeat(ID_TIME_LENGTH), ID_BASE);
const maxRandom = parseInt('1' + '0'.repeat(ID_RANDOM_LENGTH), ID_BASE);

const padLeft = (str: string, char: string, minLength: number) =>
  str.length >= minLength ? str
    : (char.repeat(minLength - str.length) + str);

const generateNodeId = () =>
  padLeft((new Date().valueOf() % maxTimestamp).toString(ID_BASE), '0', ID_TIME_LENGTH) +
  padLeft(Math.floor(Math.random() * maxRandom).toString(ID_BASE), '0', ID_RANDOM_LENGTH);

export default generateNodeId;
