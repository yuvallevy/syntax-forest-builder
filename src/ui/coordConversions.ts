import { PlotCoords } from '../core/types';
import ClientCoords from './ClientCoords';

// Calculate plot coords from client coords & vice versa, when they're different (when zoom and/or pan are implemented)

export const plotCoordsToClientCoords = (plotCoords: PlotCoords): ClientCoords => ({
  clientX: plotCoords.plotX,
  clientY: plotCoords.plotY,
});

export const clientCoordsToPlotCoords = (clientCoords: ClientCoords): PlotCoords => ({
  plotX: clientCoords.clientX,
  plotY: clientCoords.clientY,
});
