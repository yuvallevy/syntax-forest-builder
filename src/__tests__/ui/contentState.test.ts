import { describe, expect, it } from 'vitest';
import { ContentAction, ContentState, contentReducer, UndoableContentState } from '../../ui/contentState';

describe('content state', () => {
  const initialState: ContentState = {
    plots: {
      'plot': {
        trees: {
          'aa': {
            nodes: {
              'a': { label: 'N', offset: { dTreeX: 0, dTreeY: 0 }, slice: [0, 4], triangle: false },
              'b': { label: 'N', offset: { dTreeX: 0, dTreeY: 0 }, slice: [5, 10], triangle: false },
            },
            offset: { dPlotX: 0, dPlotY: 0 },
            sentence: 'tree state',
          }
        }
      }
    },
  };

  const undoableInitialState: UndoableContentState = {
    current: initialState,
    undoStack: [],
    redoStack: [],
  };

  const testCases: [string, ContentAction][] = [
    [
      'inserts a node into a tree',
      { type: 'insertNode', plotId: 'plot', treeId: 'aa', newNodeId: 'c', newNode: { label: 'NP', targetChildIds: ['a', 'b'] } }
    ],
    [
      'deletes a node from a tree',
      { type: 'deleteNodes', plotId: 'plot', nodes: [{ treeId: 'aa', nodeId: 'b' }] }
    ],
    [
      'sets the label of a node',
      { type: 'setNodeLabel', plotId: 'plot', node: { treeId: 'aa', nodeId: 'a' }, newLabel: 'NP' }
    ],
    [
      'reacts to a change in the sentence',
      { type: 'setSentence', plotId: 'plot', treeId: 'aa', newSentence: 'tee state', oldSelectedSlice: [2, 2] }
    ],
    [
      'adds a tree to a plot',
      { type: 'addTree', plotId: 'plot', newTreeId: 'zz', offset: { dPlotX: 105, dPlotY: 88 } }
    ],
    [
      'removes a tree from a plot',
      { type: 'removeTree', plotId: 'plot', treeId: 'aa' }
    ],
  ];

  it.each(testCases)('%s', (_, action) => expect(contentReducer(undoableInitialState, action).current).toMatchSnapshot());
});
