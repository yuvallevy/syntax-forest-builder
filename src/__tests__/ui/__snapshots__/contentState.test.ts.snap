// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`content state > adds a tree to a plot 1`] = `
{
  "plots": {
    "plot": {
      "trees": {
        "aa": {
          "nodes": {
            "a": {
              "label": "N",
              "offset": {
                "dTreeX": 0,
                "dTreeY": 0,
              },
              "slice": [
                0,
                4,
              ],
              "triangle": false,
            },
            "b": {
              "label": "N",
              "offset": {
                "dTreeX": 0,
                "dTreeY": 0,
              },
              "slice": [
                5,
                10,
              ],
              "triangle": false,
            },
          },
          "offset": {
            "dPlotX": 0,
            "dPlotY": 0,
          },
          "sentence": "tree state",
        },
        "zz": {
          "nodes": {},
          "offset": {
            "dPlotX": 105,
            "dPlotY": 88,
          },
          "sentence": "",
        },
      },
    },
  },
}
`;

exports[`content state > deletes a node from a tree 1`] = `
{
  "plots": {
    "plot": {
      "trees": {
        "aa": {
          "nodes": {
            "a": {
              "label": "N",
              "offset": {
                "dTreeX": 0,
                "dTreeY": 0,
              },
              "slice": [
                0,
                4,
              ],
              "triangle": false,
            },
          },
          "offset": {
            "dPlotX": 0,
            "dPlotY": 0,
          },
          "sentence": "tree state",
        },
      },
    },
  },
}
`;

exports[`content state > inserts a node into a tree 1`] = `
{
  "plots": {
    "plot": {
      "trees": {
        "aa": {
          "nodes": {
            "a": {
              "label": "N",
              "offset": {
                "dTreeX": 0,
                "dTreeY": 0,
              },
              "slice": [
                0,
                4,
              ],
              "triangle": false,
            },
            "b": {
              "label": "N",
              "offset": {
                "dTreeX": 0,
                "dTreeY": 0,
              },
              "slice": [
                5,
                10,
              ],
              "triangle": false,
            },
            "c": {
              "children": [
                "a",
                "b",
              ],
              "label": "NP",
              "offset": {
                "dTreeX": 0,
                "dTreeY": 0,
              },
            },
          },
          "offset": {
            "dPlotX": 0,
            "dPlotY": 0,
          },
          "sentence": "tree state",
        },
      },
    },
  },
}
`;

exports[`content state > reacts to a change in the sentence 1`] = `
{
  "plots": {
    "plot": {
      "trees": {
        "aa": {
          "nodes": {
            "a": {
              "label": "N",
              "offset": {
                "dTreeX": 0,
                "dTreeY": 0,
              },
              "slice": [
                0,
                3,
              ],
              "triangle": false,
            },
            "b": {
              "label": "N",
              "offset": {
                "dTreeX": 0,
                "dTreeY": 0,
              },
              "slice": [
                4,
                9,
              ],
              "triangle": false,
            },
          },
          "offset": {
            "dPlotX": 0,
            "dPlotY": 0,
          },
          "sentence": "tee state",
        },
      },
    },
  },
}
`;

exports[`content state > removes a tree from a plot 1`] = `
{
  "plots": {
    "plot": {
      "trees": {},
    },
  },
}
`;

exports[`content state > sets the label of a node 1`] = `
{
  "plots": {
    "plot": {
      "trees": {
        "aa": {
          "nodes": {
            "a": {
              "label": "NP",
              "offset": {
                "dTreeX": 0,
                "dTreeY": 0,
              },
              "slice": [
                0,
                4,
              ],
              "triangle": false,
            },
            "b": {
              "label": "N",
              "offset": {
                "dTreeX": 0,
                "dTreeY": 0,
              },
              "slice": [
                5,
                10,
              ],
              "triangle": false,
            },
          },
          "offset": {
            "dPlotX": 0,
            "dPlotY": 0,
          },
          "sentence": "tree state",
        },
      },
    },
  },
}
`;

exports[`content state > sets whether a terminal node connects to its slice with a triangle 1`] = `
{
  "plots": {
    "plot": {
      "trees": {
        "aa": {
          "nodes": {
            "a": {
              "label": "N",
              "offset": {
                "dTreeX": 0,
                "dTreeY": 0,
              },
              "slice": [
                0,
                4,
              ],
              "triangle": false,
            },
            "b": {
              "label": "N",
              "offset": {
                "dTreeX": 0,
                "dTreeY": 0,
              },
              "slice": [
                5,
                10,
              ],
              "triangle": true,
            },
          },
          "offset": {
            "dPlotX": 0,
            "dPlotY": 0,
          },
          "sentence": "tree state",
        },
      },
    },
  },
}
`;
